scripts/tests/dns-failsafe.sh

```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+
+# If we set the DNS control to `systemd-resolved` but that's not available,
+# we should still boot up and allow IP / CIDR resources to work
```

Comment:
Do we print a loud warning about this?

I'm thinking the most common sad-path here is a user turns Firezone on, thinks their queries to `private.site.com` are protected, when in fact they aren't and they're none the wiser.

Reply:
Hm, maybe we don't. I opened an issue to check and fix https://github.com/firezone/firezone/issues/4461

> thinks their traffic to private.site.com is protected, when in fact they aren't and they're none the wiser.

This is a big problem in general because of our SplitDNS setup. If something goes wrong, like the user switches Wi-Fi networks, there is definitely a gap where the tunnel is down and:
- Their DNS query for `private.site.com` goes to their DHCP-assigned DNS instead of the tunnel's upstream DNS, which could be bad
- Their traffic doesn't go through the tunnel, which could be bad

These changes would be cool if possible:
- Have some TLD like `.tunneled` (I think we had `.fz`?) that browsers / OSes will never try to route over normal Internet. This depends on web specs and may or may not exist
- Separate connlib's DNS control from the tunneling so that we can maintain DNS control even if the tunnel is reconnecting or temporarily down due to any kind of network issue
- Encourage customers to use an IP allowlist or something so that subdomains meant for private resources drop public traffic

Reply:
Yeah there's not much we can do unfortunately. Training end-users to append `.fz` everytime they need to access a resource wouldn't be realistic.

> Separate connlib's DNS control from the tunneling

This is why we keep the tunnel up when switching networks (well, except on iOS) -- as long as connlib's sentinels are still there then the packets just get dropped instead of going to the actual DNS resolver.

Reply:
hm I remember a recent manual test where `curl -4 https://ifconfig.net/ip` was returning my house's IP when switching networks.

It shouldn't do that, right? This would have only been a week or two ago while doing last looks for GA.

Reply:
Maybe Windows is different, but on macOS we never bring the tunnel, so the system never sees a different resolver other than the sentinels

Reply:
Yeah but I don't think we block out the other resolvers, even if we keep ours up.

So if Windows gets a DNS query it might say "Firezone is the best resolver, but it didn't respond for 2 seconds, so I'll ask another resolver"